lib/sshkit.ex

```diff
@@ -262,17 +263,96 @@ defmodule SSHKit do
     Enum.map(context.hosts, run)
   end
 
-  # def upload(context, path, options \\ []) do
-  #   …
-  #   # resolve remote relative to context path
-  #   remote = Path.expand(Map.get(options, :as, Path.basename(path)), _)
-  #   SCP.upload(conn, path, remote, options)
-  # end
-
-  # def download(context, path, options \\ []) do
-  #   …
-  #   remote = _ # resolve remote relative to context path
-  #   local = Map.get(options, :as, Path.basename(path))
-  #   SCP.download(conn, remote, local, options)
-  # end
+  @doc ~S"""
+  Upload a file or files to the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file upload was successful.
+
+  ## Examples
+
+  Upload all files and folders in current directory to "/workspace".
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.path("/workspace")
+    |> SSHKit.upload(".", recursive: true)
+
+  ```
+
+  Upload file to different name on host
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.upload("some_file.txt", as: "other_file.txt" )
+
+  ```
+
+  """
+  def upload(context, path, options \\ []) do
+    # resolve remote relative to context path
+    # override with :as option if present
+    destination_path = if (context.path), do: context.path, else: ""
+    destination = case Keyword.get(options, :as, false) do
+      false -> path
+      override -> override
+    end
+    remote = if (Path.type(destination) == :absolute), do: destination, else: Path.join(destination_path, destination)
+
+    run = fn host ->
+      {:ok, conn} = SSH.connect(host.name, host.options)
+      SCP.upload(conn, path, remote, options)
+    end
+
+    Enum.map(context.hosts, run)
+  end
+
+  @doc ~S"""
+  Download a file or files from the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file download was successful.
+
+  ## Examples
+
+  Download all files and folders in context directory to current working directory.
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.path("/workspace")
+    |> SSHKit.upload(".", recursive: true)
+
+  ```
+
+  Download file to different local name.
+
+  ```
+  [:ok] =
+    "my.remote-host.tld"
+    |> SSHKit.context
+    |> SSHKit.download("some_file.txt", as: "other_file.txt" )
+
+  ```
+
+  """
+  def download(context, path, options \\ []) do
```

Comment:
In order to align this with the changes we proposed for `upload`, this could look somewhat along the lines of this:
```elixir
def download(context, path, options \\ []) do
  remote = build_remote_path(context, path)
  local = Keyword.get(options, :as, Path.basename(path))

  run = fn host ->
    {:ok, conn} = SSH.connect(host.name, host.options)
    res = SCP.download(conn, remote, local, options)
    SSH.close(conn)
    res
  end

  Enum.map(context.hosts, run)
en
```
It would then also make use of the previously defined `build_remote_path` function we proposed in the changes for `upload` 🎉 

Same as `upload`, it also closes the connection per host - something that for now may seem a bit tedious, but will probably be addressed when it comes to #41 ✌️ 

And for the docs, same as for `upload`, depending on whether or not we have correct assumptions on the return values of `SCP.download` 😉